src/coordination/coordinator_instance.cpp

```diff
@@ -18,80 +18,271 @@
+                          std::string_view instance_name) -> ReplicationInstance & {
+    auto instance = std::ranges::find_if(
+        coord_instance->repl_instances_,
+        [instance_name](ReplicationInstance const &instance) { return instance.InstanceName() == instance_name; });
+
+    MG_ASSERT(instance != coord_instance->repl_instances_.end(), "Instance {} not found during callback!",
+              instance_name);
+    return *instance;
+  };
+
+  replica_succ_cb_ = [find_instance](CoordinatorInstance *coord_instance, std::string_view instance_name) -> void {
+    auto lock = std::lock_guard{coord_instance->coord_instance_lock_};
+    spdlog::trace("Instance {} performing replica successful callback", instance_name);
+    find_instance(coord_instance, instance_name).OnSuccessPing();
+  };
+
+  replica_fail_cb_ = [find_instance](CoordinatorInstance *coord_instance, std::string_view instance_name) -> void {
+    auto lock = std::lock_guard{coord_instance->coord_instance_lock_};
+    spdlog::trace("Instance {} performing replica failure callback", instance_name);
+    find_instance(coord_instance, instance_name).OnFailPing();
+  };
+
+  main_succ_cb_ = [find_instance](CoordinatorInstance *coord_instance, std::string_view instance_name) -> void {
+    auto lock = std::lock_guard{coord_instance->coord_instance_lock_};
+    spdlog::trace("Instance {} performing main successful callback", instance_name);
+
+    auto &instance = find_instance(coord_instance, instance_name);
+
+    if (instance.IsAlive()) {
+      instance.OnSuccessPing();
+      return;
+    }
+
+    bool const is_latest_main = !coord_instance->ClusterHasAliveMain_();
+    if (is_latest_main) {
+      spdlog::info("Instance {} is the latest main", instance_name);
+      instance.OnSuccessPing();
+      return;
+    }
+
+    bool const demoted = instance.DemoteToReplica(coord_instance->replica_succ_cb_, coord_instance->replica_fail_cb_);
+    if (demoted) {
+      instance.OnSuccessPing();
+      spdlog::info("Instance {} demoted to replica", instance_name);
+    } else {
+      spdlog::error("Instance {} failed to become replica", instance_name);
+    }
+  };
+
+  main_fail_cb_ = [find_instance](CoordinatorInstance *coord_instance, std::string_view instance_name) -> void {
+    auto lock = std::lock_guard{coord_instance->coord_instance_lock_};
+    spdlog::trace("Instance {} performing main failure callback", instance_name);
+    find_instance(coord_instance, instance_name).OnFailPing();
+
+    if (!coord_instance->ClusterHasAliveMain_()) {
+      spdlog::info("Cluster without main instance, trying automatic failover");
+      coord_instance->TryFailover();
+    }
+  };
+}
+
+auto CoordinatorInstance::ClusterHasAliveMain_() const -> bool {
+  auto const alive_main = [](ReplicationInstance const &instance) { return instance.IsMain() && instance.IsAlive(); };
+  return std::ranges::any_of(repl_instances_, alive_main);
+}
+
+auto CoordinatorInstance::TryFailover() -> void {
```

Comment:
Seems to have very similar logic to `SetReplicationInstanceToMain` can we unify it in some way?

Reply:
Yes, nice catch I will see what can be done